File: py/src/apptheory/cloudfront.py
from __future__ import annotations

from typing import Any

from apptheory.util import canonicalize_headers


def origin_url(headers: dict[str, Any] | None) -> str:
    h = canonicalize_headers(headers)
    forwarded = _parse_forwarded(_first_header_value(h, "forwarded"))

    host = _first_header_value(h, "x-forwarded-host") or forwarded.get("host") or _first_header_value(h, "host")
    host = _first_comma_token(host).strip()
    if not host:
        return ""

    proto = (
        _first_header_value(h, "cloudfront-forwarded-proto")
        or _first_header_value(h, "x-forwarded-proto")
        or forwarded.get("proto", "")
    )
    proto = _first_comma_token(proto).strip().lower()
    if not proto:
        proto = "https"

    return f"{proto}://{host}"


def client_ip(headers: dict[str, Any] | None) -> str:
    h = canonicalize_headers(headers)

    viewer = _first_header_value(h, "cloudfront-viewer-address")
    if viewer:
        ip = _parse_cloudfront_viewer_address(viewer)
        if ip:
            return ip

    xff = _first_header_value(h, "x-forwarded-for")
    if xff:
        ip = _first_comma_token(xff).strip()
        if ip:
            return ip

    return ""


def _first_header_value(headers: dict[str, list[str]], key: str) -> str:
    values = headers.get(str(key or "").strip().lower(), [])
    return str(values[0]) if values else ""


def _first_comma_token(value: Any) -> str:
    return str(value or "").split(",", 1)[0]


def _parse_forwarded(value: Any) -> dict[str, str]:
    raw = str(value or "").strip()
    if not raw:
        return {}

    first = raw.split(",", 1)[0]
    out: dict[str, str] = {}
    for part in first.split(";"):
        part = part.strip()
        if "=" not in part:
            continue
        key, val = part.split("=", 1)
        key = key.strip().lower()
        val = val.strip().strip("\"")
        if key in {"proto", "host"} and key not in out and val:
            out[key] = val
    return out


def _parse_cloudfront_viewer_address(value: Any) -> str:
    raw = str(value or "").strip().strip("\"")
    if not raw:
        return ""

    if raw.startswith("[") and "]" in raw:
        return raw.split("]", 1)[0].lstrip("[").strip()

    idx = raw.rfind(":")
    if idx <= 0:
        return raw
    ip_part = raw[:idx].strip()
    port_part = raw[idx + 1 :].strip()
    if not ip_part or not port_part:
        return raw
    if not port_part.isdigit():
        return raw
    return ip_part

File: runtime/cloudfront.go
package apptheory

import (
	"strings"
	"unicode"
)

// OriginURL reconstructs the canonical origin URL (scheme + host) from edge-copied original-host
// headers first, then falls back to generic forwarded headers when needed.
func OriginURL(headers map[string][]string) string {
	headers = canonicalizeHeaders(headers)

	forwardedProto, _ := parseForwardedHeader(firstHeaderValue(headers, "forwarded"))

	host := originalHostFromCanonicalizedHeaders(headers)
	if host == "" {
		return ""
	}

	proto := firstHeaderValue(headers, "cloudfront-forwarded-proto")
	if proto == "" {
		proto = firstHeaderValue(headers, "x-forwarded-proto")
	}
	if proto == "" {
		proto = forwardedProto
	}
	proto = strings.ToLower(strings.TrimSpace(firstCommaToken(proto)))
	if proto == "" {
		proto = "https"
	}

	return proto + "://" + host
}

// OriginalHost extracts the best available viewer-facing host without depending on raw Host forwarding.
func OriginalHost(headers map[string][]string) string {
	return originalHostFromCanonicalizedHeaders(canonicalizeHeaders(headers))
}

// OriginalURI returns the viewer URI captured by the edge contract when available.
func OriginalURI(headers map[string][]string) string {
	return originalURIFromCanonicalizedHeaders(canonicalizeHeaders(headers))
}

// ClientIP extracts a stable client IP address from CloudFront and generic forwarded headers.
func ClientIP(headers map[string][]string) string {
	headers = canonicalizeHeaders(headers)

	if value := firstHeaderValue(headers, "cloudfront-viewer-address"); value != "" {
		if ip := parseCloudFrontViewerAddress(value); ip != "" {
			return ip
		}
	}

	if value := firstHeaderValue(headers, "x-forwarded-for"); value != "" {
		ip := strings.TrimSpace(firstCommaToken(value))
		if ip != "" {
			return ip
		}
	}

	return ""
}

func firstCommaToken(value string) string {
	if idx := strings.Index(value, ","); idx >= 0 {
		return value[:idx]
	}
	return value
}

func originalHostFromCanonicalizedHeaders(headers map[string][]string) string {
	_, forwardedHost := parseForwardedHeader(firstHeaderValue(headers, "forwarded"))

	host := firstHeaderValue(headers, "x-apptheory-original-host")
	if host == "" {
		host = firstHeaderValue(headers, "x-facetheory-original-host")
	}
	if host == "" {
		host = firstHeaderValue(headers, "x-forwarded-host")
	}
	if host == "" {
		host = forwardedHost
	}
	if host == "" {
		host = firstHeaderValue(headers, "host")
	}
	host = firstCommaToken(host)
	return strings.TrimSpace(host)
}

func originalURIFromCanonicalizedHeaders(headers map[string][]string) string {
	value := firstHeaderValue(headers, "x-apptheory-original-uri")
	if value == "" {
		value = firstHeaderValue(headers, "x-facetheory-original-uri")
	}
	return strings.TrimSpace(value)
}

func parseForwardedHeader(value string) (proto string, host string) {
	value = strings.TrimSpace(value)
	if value == "" {
		return "", ""
	}

	first := value
	if idx := strings.Index(first, ","); idx >= 0 {
		first = first[:idx]
	}

	parts := strings.Split(first, ";")
	for _, part := range parts {
		kv := strings.SplitN(strings.TrimSpace(part), "=", 2)
		if len(kv) != 2 {
			continue
		}
		key := strings.ToLower(strings.TrimSpace(kv[0]))
		val := strings.TrimSpace(kv[1])
		val = strings.Trim(val, "\"")
		switch key {
		case "proto":
			if proto == "" {
				proto = val
			}
		case "host":
			if host == "" {
				host = val
			}
		}
	}
	return proto, host
}

func parseCloudFrontViewerAddress(value string) string {
	value = strings.TrimSpace(value)
	value = strings.Trim(value, "\"")
	if value == "" {
		return ""
	}

	if strings.HasPrefix(value, "[") {
		if idx := strings.Index(value, "]"); idx > 1 {
			return strings.TrimSpace(value[1:idx])
		}
	}

	lastColon := strings.LastIndex(value, ":")
	if lastColon <= 0 {
		return value
	}

	ipPart := strings.TrimSpace(value[:lastColon])
	portPart := strings.TrimSpace(value[lastColon+1:])
	if ipPart == "" || portPart == "" {
		return value
	}

	for _, r := range portPart {
		if !unicode.IsDigit(r) {
			return value
		}
	}

	return ipPart
}
